File: leaderboard-service/src/cleanup.js
// Releases name claims that were reserved but never played, so a
// squatted name recycles back into the pool instead of being lost
// forever. Replaces class-claim-cleanup.php's WP-Cron hourly job — this
// service stays alive as a long-running process, so a scheduled task
// needs no separate cron mechanism the way a stateless function platform
// would. Real rows (score > 0) are never touched, no matter how old — see
// "Data is retained indefinitely" in DATABASE.md.
const { buildLeaderboardKeys } = require("./keys");

const GRACE_SECONDS = 172800; // 48 hours, same grace window as before

// There's a window between this scan and the deletion below where a
// player can legitimately submit — submit.lua ZREMs their token out of
// :unplayed the instant a real score lands, but that can happen after
// this ZRANGEBYSCORE already returned their token as "stale". Deleting
// unconditionally here would race a real submit and destroy a live
// player. releaseUnplayedClaim (scripts/cleanup.lua) re-checks
// atomically, immediately before deleting, that the token is still
// unplayed — so a submit that lands mid-sweep always wins.
async function cleanupUnplayedClaims(redis, gameKey) {
  const K = buildLeaderboardKeys(gameKey);
  const cutoff = Math.floor(Date.now() / 1000) - GRACE_SECONDS;
  const staleTokens = await redis.zrangebyscore(K.unplayed, 0, cutoff);

  let released = 0;
  for (const token of staleTokens) {
    const result = await redis.releaseUnplayedClaim(
      K.unplayed, K.names, K.player(token), K.board,
      token, `wplb:${gameKey}:name:`
    );
    if (result === 1) released++;
  }

  if (released) {
    console.log(`[cleanup] released ${released} unplayed claim(s)`);
  }

  return released;
}

module.exports = { GRACE_SECONDS, cleanupUnplayedClaims };

File: leaderboard-service/scripts/cleanup.lua
-- KEYS[1] unplayed_zset ("wplb:{game}:unplayed")
-- KEYS[2] names_set     ("wplb:{game}:names")
-- KEYS[3] player_key    ("wplb:{game}:player:{token}")
-- KEYS[4] board_zset    ("wplb:{game}:board")
-- ARGV[1] token
-- ARGV[2] name_key_prefix ("wplb:{game}:name:")
--
-- Re-checks two conditions immediately before deleting anything, closing
-- the race where /submit records a real score between cleanup's initial
-- ZRANGEBYSCORE scan (cleanup.js) and this deletion actually running:
-- (1) the token must still be present in :unplayed — submit.lua ZREMs it
-- the instant a real score lands — and (2) the player's stored score must
-- still be exactly 0, as a second independent guard. Either condition
-- failing means a legitimate submit won the race; this is then a no-op
-- and returns 0 rather than deleting a live player.
if not redis.call('ZSCORE', KEYS[1], ARGV[1]) then
  return 0
end

local score = redis.call('HGET', KEYS[3], 'score')
if score ~= '0' then
  return 0
end

local playerName = redis.call('HGET', KEYS[3], 'player_name')
if playerName then
  redis.call('SREM', KEYS[2], playerName)
  redis.call('DEL', ARGV[2] .. playerName)
end
redis.call('DEL', KEYS[3])
redis.call('ZREM', KEYS[4], ARGV[1])
redis.call('ZREM', KEYS[1], ARGV[1])
return 1
